src/config-file.c

```diff
@@ -34,53 +34,86 @@ static const gint DEFAULT_TIMEOUT         = 60;     // 1 min.
 static const gint DEFAULT_RETRY_WAIT      = 5 * 60; // 5 min.
 static const gboolean DEFAULT_SSL         = TRUE;
 static const gboolean DEFAULT_SSL_VERIFY  = TRUE;
-static const gchar * DEFAULT_LOG_LEVEL    = "message";
+static const gchar* DEFAULT_LOG_LEVEL     = "message";
 
-void config_file_free(struct config *config)
+/**
+ * @brief Get string value from key_file for key in group, optional default_value can be specified
+ * that will be used in case key is not found in group.
+ *
+ * @param[in]  key_file      GKeyFile to look value up
+ * @param[in]  group         A group name
+ * @param[in]  key           A key
+ * @param[out] value         Output string value
+ * @param[in]  default_value String value to return in case no value found, or NULL (not found
+ *                           leads to error)
+ * @param[out] error         Error
+ * @return TRUE if found, TRUE if not found and default_value given, FALSE otherwise (error is set)
+ */
+static gboolean get_key_string(GKeyFile *key_file, const gchar *group, const gchar *key,
+                               gchar **value, const gchar *default_value, GError **error)
 {
-        g_free(config->hawkbit_server);
-        g_free(config->controller_id);
-        g_free(config->tenant_id);
-        g_free(config->auth_token);
-        g_free(config->gateway_token);
-        g_free(config->bundle_download_location);
-        g_hash_table_destroy(config->device);
-}
+        g_autofree gchar *val = NULL;
 
-static gboolean get_key_string(GKeyFile *key_file, const gchar* group, const gchar* key, gchar** value, const gchar* default_value, GError **error)
-{
-        gchar *val = NULL;
-        val = g_key_file_get_string(key_file, group, key, NULL);
-        if (val == NULL) {
-                if (default_value != NULL) {
+        g_return_val_if_fail(key_file, FALSE);
+        g_return_val_if_fail(group, FALSE);
+        g_return_val_if_fail(key, FALSE);
+        g_return_val_if_fail(value, FALSE);
+        g_return_val_if_fail(value && *value == NULL, FALSE);
```

Comment:
just saw that this line now duplicates the check from the line above, so please drop the original check line.

Reply:
Fixed.

Reply:
:+1: